lib/posticle/src/lib.rs

```diff
@@ -40,72 +56,131 @@ impl Entity {
     }
+
+    /// Create an Entity of `EntityKind::Hashtag` from parser result.
+    pub fn from_hashtag(pair: Pair<Rule>) -> Option<Self> {
+        let span = pair.as_span();
+        let start = span.start_pos().pos();
+        let end = span.end_pos().pos();
+        let pairs = pair.into_inner();
+
+        for pair in pairs {
+            match pair.as_rule() {
+                Rule::hashtag_name => {
+                    if validate_hashtag_name(pair.as_str()) {
+                        return Some(Entity {
+                            kind:  EntityKind::Hashtag,
+                            range: (start, end),
+                        });
+                    }
+                },
+                _ => unreachable!(),
+            }
+        }
+
+        None
+    }
+
+    /// Create an Entity of `EntityKind::Mention` from parser result.
+    pub fn from_mention(pair: Pair<Rule>) -> Option<Self> {
+        let span = pair.as_span();
+        let start = span.start_pos().pos();
+        let end = span.end_pos().pos();
+        let pairs = pair.into_inner();
+        let mut username = "";
+        let mut domain = "";
+        let mut with_domain = false;
+
+        for pair in pairs {
+            match pair.as_rule() {
+                Rule::mention_username => {
+                    username = pair.as_str();
+                },
+                Rule::mention_domain => {
+                    with_domain = true;
+                    domain = pair.as_str();
+                },
+                _ => unreachable!(),
+            }
+        }
+
+        if validate_mention_username(&username) {
+            if with_domain {
+                if validate_mention_domain(&domain) {
+                    Some(Entity {
+                        kind:  EntityKind::Mention(username.to_string(), Some(domain.to_string())),
+                        range: (start, end),
+                    })
+                } else {
+                    None
+                }
+            } else {
+                Some(Entity {
+                    kind:  EntityKind::Mention(username.to_string(), None),
+                    range: (start, end),
+                })
+            }
+        } else {
+            None
+        }
+    }
+
+    /// Create an Entity of `EntityKind::Url` from parser result.
+    pub fn from_url(pair: Pair<Rule>) -> Option<Self> {
+        let span = pair.as_span();
+        let start = span.start_pos().pos();
+        let end = span.end_pos().pos();
+
+        if validator::validate_url(pair.as_str()) {
+            Some(Entity {
+                kind:  EntityKind::Url,
+                range: (start, end),
+            })
+        } else {
+            None
+        }
+    }
 }
 
 /// Given `text`, extract all [Entities](Entity)
 pub fn entities(text: &str) -> Vec<Entity> {
-    if text.is_empty() {
-        return Vec::new();
-    }
+    let pairs = Grammar::parse(Rule::post, text).unwrap_or_else(|e| panic!("{}", e));
```

Comment:
We should probably pass the potential parse error up out of entities here?

Reply:
In our current scenario were we are simply searching a string for tokens and then returning what we find, would it be better to return a `Result` or simply an empty `Vec`?

I only ask as every change I've made so far is using the same public interface so I'm a bit loath to change it just yet.

For now I'll go the empty `Vec` route, even if we end up changing it tomorrow.